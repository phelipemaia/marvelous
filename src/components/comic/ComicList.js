import React from 'react';
import { FlatList, StyleSheet  } from 'react-native';
import ComicElement from './ComicElement';
import CacheRequest from '../../util/RequestCache';

export default class ComicList extends React.Component {
  constructor(props) {
    super(props);

    this.state = { loading: true, comics: [], offset: 0, limit: 10 };
  }

  componentWillMount() {
    this.makeRemoteRequest();
  }

  makeRemoteRequest = (discardCache = false) => {
    let url = `https://gateway.marvel.com:443/v1/public/comics?offset=${this.state.offset}&limit=${this.state.limit}`;

    CacheRequest.get('comics', url, this.state.offset, discardCache)
      .then(result => {
        let offset = result.offset;
        if (offset) {
          this.setState({comics: this.state.offset === 0 ? result : [...this.state.comics, ...result], offset: offset})
        } else {
          this.setState({comics: this.state.offset === 0 ? result : [...this.state.comics, ...result]})
        }
      })
      .catch(error => {
        console.error('Error on comics request ', error)
      });
  }

  handleLoadMore = (info) => {
    if (info.distanceFromEnd > 0) {
      this.setState({
          offset: this.state.offset + 10
        },
        () => {
          this.makeRemoteRequest();
        });
    }
  }

  render() {
    const { containerStyle } = styles;
    return (
      <FlatList data={this.state.comics} renderItem={({ item }) => (
        <ComicElement key={item.id} comic={item}></ComicElement>
      )} horizontal={true} showsHorizontalScrollIndicator={false} showsVerticalScrollIndicator={false}
                style={containerStyle} onEndReached={this.handleLoadMore} onEndReachedThreshold={0.5}>
      </FlatList>
    );
  }
}

const styles = StyleSheet.create({
  containerStyle: {
    flexDirection: 'row'
  }
});