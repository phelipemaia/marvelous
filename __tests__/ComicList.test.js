import React from 'react';
import ComicList from '../src/components/comic/ComicList';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <ComicList/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});