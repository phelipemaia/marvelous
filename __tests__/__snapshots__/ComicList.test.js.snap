// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.5}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onViewableItemsChanged={undefined}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  showsHorizontalScrollIndicator={false}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flexDirection": "row",
    }
  }
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;
